### PR TITLE
Makes the reference tensor have the complete history of transformations V2

### PR DESCRIPTION
**Pull request by csarofeen**
Respin of https://github.com/csarofeen/pytorch/pull/1509 using https://github.com/csarofeen/pytorch/pull/1535

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/index_reference_replay.cpp [hunk 1]
```diff
@@ -177,7 +191,7 @@ TensorDomain* IndexReferenceReplay::computeReplay() {
       continue;
     }
 
-    if (!visited.emplace(expr).second) {
+    if (!visisted_exprs.emplace(expr).second) {
```
Comment (rdspring1) [inline torch/csrc/jit/codegen/cuda/index_reference_replay.cpp]: nit: `visited_exprs`

File: torch/csrc/jit/codegen/cuda/compute_at_map.h [hunk 2]
```diff
@@ -135,6 +160,13 @@ class TORCH_CUDA_CU_API ComputeAtMap {
   // must be present. The Index map can come in through a GPU Lower object,
   // otherwise it will be generated if using Parallel or Loop mapping modes.
   std::unordered_map<IterDomain*, std::pair<int, int>> concrete_id_count_map_;
+
+  // Track all domains that are the result of a view operation and consumed in a
+  // subsequent expression. When resolving concrete IDs we want to set these
+  // domains to count of concrete domains = 1 so they won't be favored over post
+  // view iteration domains when resolving the parallel and loop map concrete
+  // IDs.
+  std::unordered_set<IterDomain*> rfactor_concrete_count_reset_domains;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/compute_at_map.h]: nit: rename to `..._`.

**Review by naoyam — APPROVED** [review]
This makes sense. I thought the root domain of a reference would need to match with the root domain of an indexed tensor, but that's actually not necessary. As long as we could map something from the reference to the indexed tensor, and the reference tensor could compute the indices of its root domain, that should be sufficient.

To be more specific, in the `FusionViewConcreteDomain3` test, there are two paths reaching into the final output, one coming from `tv0` of `[14, 12, 8, 100]` and another path coming from `tv2` of `[14, 100, 96]`. For both paths, the reference tensor has a root domain of `[12, 8, 100, 14]`. I thought this wouldn't work with the `tv2` path, but their leaf domains do match, so `tv2` can compute its root indices by traversing its expression history back to the root.

**Comment by csarofeen** [comment]
I will merge this PR, but as we've discussed I'm pretty confident it's not actually sufficient. Though the reference tensor has a complete history of everything we need to index multiple paths, index compute won't actually traverse all history, it will only traverse the history of the leaf nodes of the reference. This means it could miss rfactor domains we need it to traverse into in more complex situations.

**Pull request merged**